README.md
# Braathwaate FrontAccounting v2.4

This is a lightly modified fork of the
[Ap.Muthu FrontAccounting v2.4 github repo](https://github.com/apmuthu/frontac24.git,)
which is a current stable version of vanilla FrontAccounting (FA).
I chose to base my modifications off this fork,
but the
[Official GitHub Mirror Repo](https://github.com/FrontAccountingERP/FA/tree/master)
would also have been a good choice.

The "wwh" branch is the main repository.
The "proposed" branches are obsolete pull requests,
but are retained because
these pull requests were not integrated into the FA core.

The braathwaate modifications are mostly usability enhancements
designed to reduce the amount of clicking and scrolling
although there are also bug fixes and feature enhancements.

Following is a list of modifications
between the braathwaate fork (BF) and FA.

## wwh customizations

## Quick entry inactive support

In FA, the list of Quick entries accumulate ad infinitum.

BF supports marking the Quick entry inactive so it does not
show up in the payment list.

## Allow bank account payments on supplier direct invoice
FA restricts payment on a supplier direct invoice
to cash accounts.
For other forms of payments,
one must select "Delayed"
and enter the payment separately.

BF allows any bank account to be used for the payment,
such as a credit card.
## Allow bank accounts in POS
FA restricts payment on a customer direct invoice
to cash accounts,
because the POS setup is restricted to cash accounts.

POS systems often allow direct deposit into bank accounts,
so BF does not restrict the POS "cash sale" to cash accounts
## apportion service items; bug fixes
This mod gives the user the option on a PO to apportion service items (freight, setup charges, discounts, etc) into prices of non-service items on the PO without changing the PO order total. By apportioning these costs into item prices, they are accounted for in inventory and thus accounted for in COGS only when then items are sold. Three methods of apportionment are provided: by item quantity, line item, and line total.

additional bug fixes:

    Viewing order totals on POs with items with decimal precision greater than user decimals could be wrong because of rounding. Changes in the code reflect the method used to generate the PO, by rounding each line item to user decimals, then adding. Prior code added then rounded.

    php.ini precision could cause disabling of FA PO decimal precision for items with precise Conversion Factors. Fix was to use php.ini precision rather than disable the feature.

    credit memos for tax exempt items do not have tax gl account, and code generated a PHP warning. This did not cause any problem other than the PHP warning, because G/L was created on the remaining total.

## ajax browser sync
In FA, when a user modifies selection criteria on a FA page and AJAX updates the page, then the user clicks to another page such as via the editor, the browser back button (or FA back link) does not return to the page as the user modified it. This forces the user to reenter the selection criteria and hit the Show/Search button again.

This mod updates the browser page history when AJAX updates the page so that the back button returns to the FA page as the user modified it.
## keep reconcile on bank trans mod
When a bank transaction is modified, FA unreconciles it, even if
just the memo or date was changed.

BF retains the reconciled status.

## new combo4 autocomplete
This option (config.php: $search_autocomplete = true)
allows finding items, customers and suppliers quickly.
The configuration parameter "search_autocomplete" when set to true
changes the behavior of the Search Item List, Customer List, and
Supplier List checkboxes in the Company Setup.

When this option is set to "true",
an autocomplete list is updated after every character
of the search text is entered into the search box.
Thus, once the desired search item appears on the autocomplete list,
the user can make just one click to select the search item.

More keystrokes are required in in FA to search.
After the user enters the search text,
the user has to hit the enter key,
use the pull-down list to find the item,
and click on the item to select it.
And if the search term did not find the desired item,
the user would have to repeat the process, taking more time.

BF also makes the dropdown list work exactly the same
even when the checkboxes are checked;
the checkbox just adds the ability to use the space key
to open a search box.
Otherwise the drop lists work normally.
So really there isn't any need for these checkboxes anymore.

For stock items,
the BF autocomplete option searches by item code or description.
A search is initiated with either the space bar or control-b.
The latter is useful for barcode readers programmed to
initiate a barcode with control-b.

## improve reports
FA displays an empty PDF if there is no data to display.

BF displays a message.
## Add delete link, direct link to void trans
Voiding a transaction is time-consuming and error prone in FA because
the user must provide a transaction number.

BF adds a red X icon to transactions.
Deleting transaction becomes as simple as clicking the red X
rather than having to navigate to Setup->Void.
## Add dimension to payments
FA does not dimension payments,
which causes a problem for cash basis accounting.

BF properly dimensions payments.
## allow negative quantity and amount on invoices

FA does not support invoices or payments
with negative quantities or amounts for customer refunds.
It expects that a credit memo be issued.
Nor does FA does not have the ability
to create a direct credit memo, nor does is it able to enter
purchased items and returned items on the same invoice.

BF allows this.
Besides code changes, this required altering
the "amt" field in table cust_allocations from "double unsigned"
to "double":

    mysql alter table cust_allocations modify amt double;

This has no adverse effect, and vanilla FA will also work
correctly on the altered table.

Note that negative quantities on an invoice
are returned to stock in BF.  If the customer returned an
item that cannot be returned to stock, then an inventory
adjustment will need to be made.  This is a manual operation.

## display balance in balance column and add debit/credit total

FA G/L inquiry is confusing because it displays the ending balance
under debit/credit.

BF displays the debit/credit in the proper columns
and the balance under the balance column.

## customer inquiry: display zero invoices; filter on sales type

BF displays zero invoices.
A zero invoice is often used when supplying
a customer with free product for marketing purposes.

## rename statement to register

FA uses the confusing term Bank Statement on Bank Account Inquiry
and on reports.

BF renames Bank Statement to Bank Account Inquiry and changes
reports to use Bank Register,
because data displayed is from FA
(all transactions entered)
and not from a bank statement
(all transactions processed by the bank).

## check_negative_stock allow for rounding on imprecise amounts

Decimal precise stock quantities
cause problems in FA because it is impossible to get them zeroed out.

BF assumes stock is zero when it falls below the precision
specified in Setup.

## g/l list description first to enable search by description

## BUG FIX: supplier inquiry GRN items filter by date

## FEATURE: use ajax for numeric calculations (part 1: purchasing)

FA is not a very responsive calculator or spreadsheet,
requiring the user to click an update button to do any math.

BF enables ajax on numeric text fields so that totals
are calculated as the user enters data in fields.

## add editor/delete funcs to reconcile

## allow user_transaction_days to be in future (negative)

BF allows a negative user_transaction_days in Setup so that
pages will display out into the future.

## supplier and customer name links on inquiry pages

BF makes supplier and customer names a html link
to their respective information pages
on the banking inquiry pages.
This allows one to quickly check or modify supplier/customer
information from the inquiry pages.
BF also adds the magnifying glass search icon next to these names
and this links to the supplier/customer inquiry pages.

## allow any user to edit another user transaction (disable user reference check)
BF allows a user to edit other user's transactions.

## move due date on invoice from header to footer and display only if delayed

FA displays a due date on direct supplier invoices.

BF only displays the due date if the transaction is delayed.
That is because it is meaningless if the transaction is being processed
immediately.

## remove more zero entries from Customer Balance report

The FA Customer Balance report still prints some zero entries
when no-zeros is selected.

BF fixes this problem.

## ajax enable account pull down menu on gl and bank inquiry


core/CHANGELOG.txt
2018-10-11 Janusz Dobrowolski
Work Order voided reference reusage, costing and reporting issues fixed
M	manufacturing/includes/db/work_orders_db.inc
M	reporting/includes/reports_classes.inc
M	reporting/rep409.php
M	manufacturing/includes/db/work_order_costing_db.inc

2018-07-12 Joe Hunt
Stock goes negative even if negative inventory is turned off. Fix by @kvvaradha
M	admin/void_transaction.php

2018-07-12 Joe Hunt
Fiscal year delete leads to undefined variable $total. Fixed.
M	admin/db/fiscalyears_db.inc

2018-06-18 Joe Hunt
AP Surf Set in demo data rerun
M	sql/en_US-demo.sql

2018-06-16 Joe Hunt
Report 102 Aged Customer Analysis SQL fix
M	purchasing/includes/db/suppliers_db.inc
 
2018-06-12 Joe Hunt
Wrong standard cost for AP Surf Set with rerun for wrong type
M	sql/en_US-demo.sql

2018-06-04 Joe Hunt
Work Order Report rep402 option for GL rows and re-run
M	reporting/rep402.php
M	reporting/reports_main.php

2018-06-01 Janusz Dobrowolski
Synched all install languages with upstream
M	install/lang/ar_EG/LC_MESSAGES/ar_EG.po	ar_EG.po
M	install/lang/da_DK/LC_MESSAGES/da_DK.po	da_DK.po
M	install/lang/de_DE/LC_MESSAGES/de_DE.po	de_DE.po
M	install/lang/el_GR/LC_MESSAGES/el_GR.po	el_GR.po
M	install/lang/es_MX/LC_MESSAGES/es_MX.po	es_MX.po
M	install/lang/fr_FR/LC_MESSAGES/fr_FR.po	fr_FR.po
M	install/lang/hi_IN/LC_MESSAGES/hi_IN.po	hi_IN.po
M	install/lang/id_ID/LC_MESSAGES/id_ID.po	id_ID.po
M	install/lang/it_IT/LC_MESSAGES/it_IT.po	it_IT.po
M	install/lang/ka_GE/LC_MESSAGES/ka_GE.po	ka_GE.po
M	install/lang/nl_BE/LC_MESSAGES/nl_BE.po	nl_BE.po
M	install/lang/pl_PL/LC_MESSAGES/pl_PL.mo	pl_PL.mo
M	install/lang/pl_PL/LC_MESSAGES/pl_PL.po	pl_PL.po
M	install/lang/pt_BR/LC_MESSAGES/pt_BR.po	pt_BR.po
M	install/lang/pt_PT/LC_MESSAGES/pt_PT.po	pt_PT.po
M	install/lang/sv_SE/LC_MESSAGES/sv_SE.po	sv_SE.po
M	install/lang/ta_IN/LC_MESSAGES/ta_IN.po	ta_IN.po
M	install/lang/vi_VN/LC_MESSAGES/vi_VN.po	vi_VN.po
M	install/lang/zh_CN/LC_MESSAGES/zh_CN.po	zh_CN.po

2018-06-01 Janusz Dobrowolski
0004461 Installer: added info about needed user database permissions. (update install langs separately)
M	install/index.php
M	install/lang/new_language_template/LC_MESSAGES/empty.po

2018-05-31 Joe Hunt
0004468: @braathwaate: Cannot modify a funds transfer twice.
M	gl/includes/db/gl_db_banking.inc

2018-05-31 Joe Hunt
0004476: Our Order No is always 0 on customer payment receipt. Fixed by empty column.
M	reporting/includes/doctext.inc

2018-05-31 Joe Hunt
0003317: "Forty" is misspelled on remittances, receipts. Fixed.
M	includes/ui/ui_view.inc

2018-05-25 Joe Hunt
Currency Convertor Google URL update
M	gl/includes/db/gl_db_rates.inc

2018-05-09 Joe Hunt
Adding Attachments to Dimension
M	dimensions/dimension_entry.php

2018-05-08 Joe Hunt
The function get_customer_trans_details now includes itemPrice ex tax in SQL.
M	sales/includes/db/cust_trans_details_db.inc

2018-04-29 Joe Hunt
0004161: g/l inquiry person/item displays number instead of person & Rerun.
M	gl/includes/db/gl_db_trans.inc

2018-04-18 Joe Hunt
Fixed a php 7.2 bug in db_pager.inc.
M	includes/db_pager.inc

2018-04-18 Joe Hunt
Typo in report Customer Detail Listing
M	reporting/rep103.php

2018-04-17 Janusz Dobrowolski
Removed obsolete column 'Received' from direct purchase document entry forms.
M	purchasing/includes/ui/po_ui.inc

2018-04-16 Joe Hunt
Standard Charts and upgrades for new sys_prefs flag ref_no_auto_increase
M	sql/en_US-demo.sql
M	sql/en_US-new.sql
M	sql/alter2.4.php

2018-04-16 Joe Hunt
Auto Increase of Document References (company setup)
M	admin/company_preferences.php
M	sales/includes/cart_class.inc

2018-04-09 Joe Hunt
Fixed Session error in PHP 7.2.
M	includes/session.inc

2018-04-09 Joe Hunt
report_classes.inc - Parameter must be an array fixed.
M	reporting/includes/reports_classes.inc

2018-04-02 Joe Hunt
Fixed the summation of the outstanding open balance in Customer Balance Report.
M	reporting/rep101.php

2018-03-27 Joe Hunt
Update Journal Entry SQL query fixed.
M	gl/includes/db/gl_journal.inc

2018-03-17 Joe Hunt
FA 2.4.4
M	version.php

2018-03-17 Janusz Dobrowolski
Moved installed languages to core, updated others
M	install/lang/ar_EG/LC_MESSAGES/ar_EG.mo
M	install/lang/ar_EG/LC_MESSAGES/ar_EG.po
M	install/lang/de_DE/LC_MESSAGES/de_DE.mo
M	install/lang/de_DE/LC_MESSAGES/de_DE.po
M	install/lang/hi_IN/LC_MESSAGES/hi_IN.mo
M	install/lang/hi_IN/LC_MESSAGES/hi_IN.po
M	install/lang/id_ID/LC_MESSAGES/id_ID.mo
M	install/lang/id_ID/LC_MESSAGES/id_ID.po

core/js/inserts.js
/**********************************************************************
    Copyright (C) FrontAccounting, LLC.
	Released under the terms of the GNU General Public License, GPL,
	as published by the Free Software Foundation, either version 3
	of the License, or (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the License here <http://www.gnu.org/licenses/gpl-3.0.html>.
***********************************************************************/
var _focus;
var _hotkeys = {
	'alt': false,	// whether is the Alt key pressed
	'list': false, // list of all elements with hotkey used recently
	'focus': -1		// currently selected list element
};

function validate(e) {
	if (e.name && (typeof _validate[e.name] == 'function'))
		return _validate[e.name](e);
	else {
		var n = e.name.indexOf('[');
		if(n!=-1) {
			var key = e.name.substring(n+1, e.name.length-1);
			if (key.length>1 && _validate[e.name.substring(0,n)])
				return _validate[e.name.substring(0,n)][key](e);
		}
	}
	return true;
}

function set_fullmode() {
	document.getElementById('ui_mode').value = 1;
	document.loginform.submit();
	return true;
}

function save_focus(e) {
  _focus = e.name||e.id;
  var h = document.getElementById('hints');
  if (h) {
	h.style.display = e.title && e.title.length ? 'inline' : 'none';
	h.innerHTML = e.title ? e.title : '';
  }
}

function _expand(tabobj) {

  var ul = tabobj.parentNode.parentNode;
  var alltabs=ul.getElementsByTagName("button");
  var frm = tabobj.form;

  if (ul.getAttribute("rel")){
	for (var i=0; i<alltabs.length; i++){
	  alltabs[i].className = "ajaxbutton"  //deselect all tabs // Review CP 2014-11 This will remove all other classes from the element.
	}
	tabobj.className = "current";
	JsHttpRequest.request(tabobj)
  }
}

//interface for selecting a tab (plus expand corresponding content)
function expandtab(tabcontentid, tabnumber) {
  var tabs = document.getElementById(tabcontentid);
 _expand(tabs.getElementsByTagName("input")[tabnumber]);
}

var _autocomplete=true;

function _set_combo_input(e) {
		e.setAttribute('_last', e.value);
		e.onblur=function() {
		  var but_name = this.name.substring(0, this.name.length-4)+'button';
		  var button = document.getElementsByName(but_name)[0];
		  var select = document.getElementsByName(this.getAttribute('rel'))[0];
		  save_focus(select);
// submit request if there is submit_on_change option set and
// search field has changed.

		  if (button && (this.value != this.getAttribute('_last'))) {
			JsHttpRequest.request(button);
		  } else if(string_contains(this.className, 'combo2')) {
				this.style.display = 'none';
				select.style.display = 'inline';
				setFocus(select);
		  }
		  return false;
		};
		e.onkeyup = function(ev) {
		    if (string_contains(this.className, 'combo4')) {
                if (!_autocomplete)
                    return;
                var div = document.getElementById(this.getAttribute('rel')+"_div");
                var ac = this.value.toUpperCase();
                b = div.getElementsByTagName("a");
                count=0;
                for (i = 0; i < b.length; i++) {
                    if ((b[i].innerHTML.toUpperCase().indexOf(ac) > -1
                        || b[i].getAttribute('data-value').toUpperCase().indexOf(ac) > -1)
                             && count < 12) {
                        b[i].style.display = "block";
                        count++;
                    } else {
                        b[i].style.display = "none";
                    }
                }
                return;
            }
			var select = document.getElementsByName(this.getAttribute('rel'))[0];
			if(select && select.selectedIndex>=0) {
			  var len = select.length;
			  var byid = string_contains(this.className, 'combo') || string_contains(this.className, 'combo3');
			  var ac = this.value.toUpperCase();
			  select.options[select.selectedIndex].selected = false;
			  var selIndex = -1;
			  for (i = 0; i < len; i++) {
				var txt = byid ? select.options[i].value : select.options[i].text;
				if (string_contains(this.className, 'combo3')) {
				  if(txt.toUpperCase().indexOf(ac) == 0) {
					selIndex=i;
				  	break;
				  }
				} else {
				  if(txt.toUpperCase().indexOf(ac) >= 0) {
					selIndex=i;
					select.options[i].removeAttribute("hidden");
				  } else
					select.options[i].setAttribute("hidden", true);
				}
			  }
			  if (selIndex != -1)
				select.options[selIndex].selected = true;
			}
		};
    	e.onkeydown = function(ev) {
	  		ev = ev||window.event;
	  		key = ev.keyCode||ev.which;
	  		if(key == 13) {
		        if(string_contains(this.className, 'combo4')) {
                    var select = document.getElementsByName(this.getAttribute('rel'))[0];
                    this.style.display = 'none';
                    select.style.display = 'inline';
                    setFocus(select);
                } else
                  this.blur();
	  		  return false;
	  		}
		}
}

function _update_box(s) {
	var byid = string_contains(s.className, 'combo') || string_contains(s.className, 'combo3');
	var rel = s.getAttribute('rel');
	var box = document.getElementsByName(rel)[0];
		if(box && s.selectedIndex>=0) {
			  var opt = s.options[s.selectedIndex];
				if(box) {
				  var old = box.value;
				  box.value = byid ? opt.value : opt.text;
				  box.setAttribute('_last', box.value);
				  return old != box.value
				}
		}
}

function _set_combo_select(e) {
		// When combo position is changed via js (eg from searchbox)
		// no onchange event is generated. To ensure proper change
		// signaling we must track selectedIndex in onblur handler.
		e.setAttribute('_last', e.selectedIndex);
		e.onblur = function() {
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
			if ((this.selectedIndex != this.getAttribute('_last'))
				||((string_contains(this.className, 'combo') || string_contains(this.className, 'combo3')) && _update_box(this))
				)
					this.onchange();
		}
		e.onchange = function() {
			var s = this;
			this.setAttribute('_last', this.selectedIndex);
			if(string_contains(s.className, 'combo') || string_contains(this.className, 'combo3'))
			    _update_box(s);
			if (validate(e) && s.selectedIndex>=0) {
				 var sname = '_'+s.name+'_update';
				 var update = document.getElementsByName(sname)[0];
				 if(update) {
                        if (typeof changeVar === "function")
                            changeVar();
					    JsHttpRequest.request(update);
				}
			}
			return true;
		}
		e.onkeydown = function(event) {
		    event = event||window.event;
		    key = event.keyCode||event.which;
		    var box = document.getElementsByName(this.getAttribute('rel'))[0];
	  		if(key == 8 || (key==37 && event.altKey)) {
				event.returnValue = false;
  			  	return false;
  			}
            // spacebar or ctrl-B (used for barcode reader) activates search
		    if (box && (key == 32 || (key == 66 && event.ctrlKey)) && (string_contains(this.className, 'combo2') || string_contains(this.className, 'combo4') )) {
                if (key == 66)
                    _autocomplete = false;  // save some time
                else
                    _autocomplete = true;
                this.style.display = 'none';
                box.style.display = 'inline';
                box.value='';
                setFocus(box);
			    return false;
			 } else {
			 	if (key == 13 && !e.length) // prevent chrome issue (blocked cursor after CR on empty selector)
			 		return false;
			 }
		}
}

var _w;

function callEditor(key) {
  var el = document.getElementsByName(editors[key][1])[0];
  if(_w) _w.close(); // this is really necessary to have window on top in FF2 :/
  var left = (screen.width - editors[key][2]) / 2;
  var top = (screen.height - editors[key][3]) / 2;
  _w = open(editors[key][0]+el.value+'&popup=1',
	  "edit","scrollbars=yes,resizable=0,width="+editors[key][2]+",height="+editors[key][3]+",left="+left+",top="+top+",screenX="+left+",screenY="+top);
  if (_w.opener == null)
	  _w.opener = self;
  editors._call = key; // store call point for passBack
  _w.focus();
}

function passBack(value) {
	var o = opener;
	if(value != false) {
		var back = o.editors[o.editors._call]; // form input bindings
		var to = o.document.getElementsByName(back[1])[0];
		if (to) {
			if (to[0] != undefined)
				to[0].value = value; // ugly hack to set selector to any value
			to.value = value;
			// update page after item selection
			o.JsHttpRequest.request('_'+to.name+'_update', to.form);
			o.setFocus(to.name);
		}
	}
	close();
}

/*
	Normalize date format using previous input value to guess missing date elements.
	Helps fast date input field change with only single or double numbers (for day or day-and-month fragments)
*/
function fix_date(date, last)
{
	var dat = last.split(user.datesep);
	var cur = date.split(user.datesep);
	var day, month, year;

// TODO: user.date as default?
// TODO: user.datesys
	if (date == "" || date == last) // should we return an empty date or should we return last?
		return date;
	if (user.datefmt == 0 || user.datefmt == 3) // set defaults
	{
		day = dat[1]; month = dat[0]; year = dat[2];
	} else if (user.datefmt == 1 || user.datefmt == 4){
		day = dat[0]; month = dat[1]; year = dat[2];
	} else {
		day = dat[2]; month = dat[1]; year = dat[0];
	}
	if (cur[1] != undefined && cur[1] != "") // day or month entered, could be string 3
	{
		if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			day = cur[1];
		else
			month = cur[1];
	}
	if (cur[0] != undefined && cur[0] != "") // day or month entered. could be string 3
	{
		if (cur[1] == undefined || cur[1] == "")
			day = cur[0];
		else if (user.datefmt == 0 || user.datefmt == 3 || ((user.datefmt == 2 || user.datefmt == 5) && (cur[2] == undefined || cur[2] == "")))
			month = cur[0];
		else if (user.datefmt == 2 || user.datefmt == 5)
			year = cur[0];
		else
			day = cur[0];
	}
	if (cur[2] != undefined && cur[2] != "") // year,
	{
		if (user.datefmt == 2 || user.datefmt == 5)
			day = cur[2];
		else
			year = cur[2];
	}
	if (user.datefmt<3) {
		if (day<10) day = '0'+parseInt(day, 10);
		if (month<10) month = '0'+parseInt(month, 10);
	}
	if (year<100) year = year<60 ? (2000+parseInt(year,10)) : (1900+parseInt(year,10));

//	console.info(day,month,year)
	if (user.datefmt == 0 || user.datefmt==3)
		return month+user.datesep+day+user.datesep+year;
	if (user.datefmt == 1 || user.datefmt==4)
		return day+user.datesep+month+user.datesep+year;
	return year+user.datesep+month+user.datesep+day;
}

/*
 Behaviour definitions
*/
var inserts = {
	'input': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
				save_focus(this);
				if (string_contains(this.className, 'combo') || string_contains(this.className, 'combo3'))
					this.select();
			};
		}
		if (string_contains(e.className, 'combo') || string_contains(e.className, 'combo2') || string_contains(e.className, 'combo3') || string_contains(e.className, 'combo4')) {
				_set_combo_input(e);
		}
		else
    		if(e.type == 'text' ) {
   	  			e.onkeydown = function(ev) {
  					ev = ev||window.event;
  					key = ev.keyCode||ev.which;
 	  				if(key == 13) {
						if(e.className == 'searchbox') {
                            e.onblur();
    						return false;
                        }
                        e.form.submit();
					}
                    return true;
	  			}
			}
	},
	'input.combo2,input.combo4,input[aspect="fallback"]':
	function(e) {
  	    // this hides search button for js enabled browsers
	    e.style.display = 'none';
	},
	'div.js_only':
	function(e) {
  	    // this shows divs for js enabled browsers only
	    e.style.display = 'block';
	},
	'button': function(e) {
            e.onclick = function(){
			if (validate(e)) {
				setTimeout(function() {	var asp = e.getAttribute('aspect');
					if (asp && asp.indexOf('download') === -1)
						set_mark((asp && ((asp.indexOf('process') !== -1) || (asp.indexOf('nonajax') !== -1))) ? 'progressbar.gif' : 'ajax-loader.gif');
				}, 100);
				return true;
			}
		},
		e.onkeydown = function(ev) {	// block unintentional page escape with 'history back' key pressed on buttons
			ev = ev||window.event;
 			key = ev.keyCode||ev.which;
	  		if(key == 8 || (key==37 && ev.altKey)) {
				ev.returnValue = false;
  			  	return false;
  			}
		}

	},
//	'.ajaxsubmit,.editbutton,.navibutton': // much slower on IE7
	'button.ajaxsubmit,input.ajaxsubmit,input.editbutton,button.editbutton,button.navibutton':
	function(e) {
			e.onclick = function() {
				if (validate(e)) {
                    if (typeof changeVar === "function")
                        changeVar();

					save_focus(e);
					var asp = e.getAttribute('aspect')
					if (asp && (asp.indexOf('process') !== -1))
						JsHttpRequest.request(this, null, 600000); // ten minutes for backup
					else
						JsHttpRequest.request(this);
				}
				return false;
			}
	},
    '.amount': function(e) {
		if(e.onblur==undefined) {
  		  e.onblur = function() {
			var dec = this.getAttribute("dec");
			price_format(this.name, get_amount(this.name), dec);
		  };
		}
	},
	'.searchbox': // emulated onchange event handling for text inputs
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off'); //must be off when calling onblur
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.setAttribute('_last_val', this.value);
					JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'.date':
		function(e) {
			e.setAttribute('_last_val', e.value);
			e.setAttribute('autocomplete', 'off');
  		  	e.onblur = function() {
				var val = this.getAttribute('_last_val');
				if (val != this.value) {
					this.value = fix_date(this.value, val);
					this.setAttribute('_last_val', this.value);
					if (e.className.match(/\bactive\b/))
						JsHttpRequest.request('_'+this.name+'_changed', this.form);
				}
			}
	},
	'button[aspect*selector], button[aspect*abort], input[aspect*selector]': function(e) {
		e.onclick = function() {
			passBack(this.getAttribute('rel'));
			return false;
		}
	},
	'button[aspect=popup]': function(e) {
		e.onclick = function() {
			if(_w) _w.close(); // this is really necessary to have window on top in FF2 :/
			 var left = (screen.width - 800)/2;
			 var top = (screen.height - 600)/2;
			  _w = open(document.location+'popup=1',
				  "edit","Scrollbars=0,resizable=0,width=800,height=600, top="+top+",left="+left+",screenX="+left+",screenY="+top);
			  if (_w.opener == null)
				  _w.opener = self;
			//  editors._call = key; // store call point for passBack
//			  _w.moveTo(50, 50);
			  _w.focus();
			return false;
		}
	},
	'select': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
  		var c = e.className;
		if (string_contains(c, 'combo') || string_contains(c, 'combo2') || string_contains(c, 'combo3') || string_contains(c, 'combo4') )
			_set_combo_select(e);
		else {
			e.onkeydown = function(ev) {	// block unintentional page escape with 'history back' key pressed on buttons
				ev = ev||window.event;
 				key = ev.keyCode||ev.which;
	  			if(key == 8 || (key=37 && ev.altKey)) {
					ev.returnValue = false;
	  			  	return false;
  				}
			}
		}
	},
	'a.printlink': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
			JsHttpRequest.request(this, null, 60000);
			return false;
		}
	},
    'a.combo4' : function(l) {
        l.onclick = function() {
            var sname = '_'+l.dataset.rel+'_update';
            var select = document.getElementsByName(l.dataset.rel)[0];
            select.value=l.dataset.value;

            var box = document.getElementsByName(select.getAttribute('rel'))[0];
            box.style.display = 'none';
            select.style.display = 'inline';
            setFocus(select);

            var update = document.getElementsByName(sname)[0];
            if(update) {
                if (typeof changeVar === "function")
                    changeVar();
                JsHttpRequest.request(update);
            }
            return false;
        }
    },
	'a.repopts_link': 	function(l) {
		l.onclick = function() {
		    save_focus(this);
		    var replinks = document.getElementsBySelector('a.repopts_link');
				for(var i in replinks)
					replinks[i].style.fontWeight = replinks[i]==this ? 'bold' : 'normal';
			JsHttpRequest.request(this, null);
			return false;
		}
	},
	'a': function(e) { // traverse menu
  		e.onkeydown = function(ev) {
			ev = ev||window.event;
			key = ev.keyCode||ev.which;
			if(key==37 || key==38 || key==39 || key==40) {
					move_focus(key, e, document.links);
					ev.returnValue = false;
					return false;
			}
		}
		// prevent unneeded transaction entry abortion
		if (e.className == 'shortcut'
		 || e.className == 'menu_option'
		 || e.className == 'menu_tab'
 		 || e.className == 'selected')
			e.onclick = function(ev) {
				if (_validate._processing
				 && _validate._modified
				 && !confirm(_validate._processing)) {
					ev.returnValue = false;
					return false;
				}
				if (_hotkeys.alt)	// ommit Chrome accesskeys
					return false;
 				window.location = e.href;
			}
	},
	'ul.ajaxtabs':	function(ul) {
	    var ulist=ul.getElementsByTagName("li");
	    for (var x=0; x<ulist.length; x++){ //loop through each LI e
		var tab=ulist[x].getElementsByTagName("button")[0];
		    var url = tab.form.action
		    tab.onclick=function(){
		    if (!_hotkeys.alt && !this.disabled)
				_expand(this);
			return false;
		    }
		}
//	    }
	},
	'textarea': function(e) {
		if(e.onfocus==undefined) {
			e.onfocus = function() {
			    save_focus(this);
			};
		}
	}
/* TODO
	'a.date_picker':  function(e) {
	    // this un-hides data picker for js enabled browsers
	    e.href = date_picker(this.getAttribute('rel'));
	    e.style.display = '';
	    e.tabindex = -1; // skip in tabbing order
	}
*/
};

function stopEv(ev) {
			if(ev.preventDefault) {
				ev.preventDefault();
				ev.stopPropagation();
			} else {
				ev.returnValue = false;
				ev.cancelBubble = true;
				window.keycode = 0;
			}
			return false;
}
/*
	Modified accesskey system. While Alt key is pressed letter keys moves
	focus to next marked link. Alt key release activates focused link.
*/
function setHotKeys() {
	document.onkeydown = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;
		if (key == 18 && !ev.ctrlKey) {	// start selection, skip Win AltGr
			_hotkeys.alt = true;
			_hotkeys.focus = -1;
			return stopEv(ev);
		}
		else if (ev.altKey && !ev.ctrlKey && ((key>47 && key<58) || (key>64 && key<91))) {
			key = String.fromCharCode(key);
			var n = _hotkeys.focus;
			var l = document.getElementsBySelector('[accesskey='+key+']');
			var cnt = l.length;
			_hotkeys.list = l;
			for (var i=0; i<cnt; i++) {
				n = (n+1)%cnt;
				// check also if the link is visible
				if (l[n].accessKey==key && (l[n].offsetWidth || l[n].offsetHeight)) {
					_hotkeys.focus = n;
	    // The timeout is needed to prevent unpredictable behaviour on IE.
					var tmp = function() {l[_hotkeys.focus].focus();};
					setTimeout(tmp, 0);
					break;
				}
			}
			return stopEv(ev);
		}
		if((ev.ctrlKey && key == 13) || key == 27) {
			_hotkeys.alt = false; // cancel link selection
			_hotkeys.focus = -1;
			ev.cancelBubble = true;
   			if(ev.stopPropagation) ev.stopPropagation();
			// activate submit/escape form
			for(var j=0; j<this.forms.length; j++) {
				var form = this.forms[j];
				for (var i=0; i<form.elements.length; i++){
					var el = form.elements[i];
					var asp = el.getAttribute('aspect');

					if (!string_contains(el.className, 'editbutton') && (asp && asp.indexOf('selector') !== -1) && (key==13 || key==27)) {
						passBack(key==13 ? el.getAttribute('rel') : false);
						ev.returnValue = false;
						return false;
					}
					if (((asp && asp.indexOf('default') !== -1) && key==13)||((asp && asp.indexOf('cancel') !== -1) && key==27)) {
						if (validate(el)) {
							if (asp.indexOf('nonajax') !== -1)
								el.click();
							else
							if (asp.indexOf('process') !== -1)
								JsHttpRequest.request(el, null, 600000);
							else
								JsHttpRequest.request(el);
						}
						ev.returnValue = false;
						return false;
					}
				}
			}
			ev.returnValue = false;
			return false;
		}
		if (editors!=='undefined' && editors[key]) {
			callEditor(key);
			return stopEv(ev); // prevent default binding
		}
		return true;
	};
	document.onkeyup = function(ev) {
		ev = ev||window.event;
		key = ev.keyCode||ev.which;

		if (_hotkeys.alt==true) {
			if (key == 18) {
				_hotkeys.alt = false;
				if (_hotkeys.focus >= 0) {
					var link = _hotkeys.list[_hotkeys.focus];
					if(link.onclick)
						link.onclick();
					else
						if (link.target=='_blank') {
							window.open(link.href,'','toolbar=no,scrollbar=no,resizable=yes,menubar=no,width=900,height=500');
							openWindow(link.href,'_blank');
						} else
							window.location = link.href;
				}
			return stopEv(ev);
			}
		}
		return true;
	}
}

Behaviour.register(inserts);

Behaviour.addLoadEvent(setFocus);
Behaviour.addLoadEvent(setHotKeys);
